### ISSUE TITLE
Mistake at cpbook-code/ch3/dp/LIS.cpp 

### ISSUE THREAD

**Issue by git-hub-tig** [stevenhalim/cpbook-code#57]
Tks your wonderful effort, great book!
But sorry some problem about the function `int LIS(int i)`
LIS is a special DP problem, need to check every element of the array. So a better approach is bottom-up that use tabulation. 
But **cpbook-code/ch3/dp/LIS.cpp** build the DP using DAG to find the single-source longest path. This method only works for the case that the last element of the array is the max, and initial `ans=1`. But not for the case like {5 4 2 3 1}. 
Because LIS is special that need to **ensure every element would be recursively 
checked!**
Additional caution, if initial using `ans=0` then, only the first element of the array is the min, the base case `i==0` would be recursively visited for every check that should be recursively visited. So using `ans=0` need to ensure the first element is the min of the arrary.

**Comment by stevenhalim** [comment]
Hi git-hub-tig,

-. top-down or bottom-up is just preference, I prefer top-down, so will remain top-down in CP4 :).
-. I now notice the bug with {5,4,3,2,1} on the very old O(n^2) that was never tested again after I use O(n log k) version all these time, thanks for catching this. I have now changed ans = 0; to ans = 1; before the recursive case to fix this case without resorting to ensure that i==0 is the minimum, i.e., LIS can now start from any index and just end at n-1 that is set to have "INF" (very large number).
-. I have tested {5,4,3,2,1,99 (need last number = "INF")} -> returns 2 of course, and now test this code with many random tc.
-. I push a newer file, pls stress test it again
